### ISSUE TITLE
Resume with shell block not working

### ISSUE THREAD

**Issue by onuryukselen** [nextflow-io/nextflow#3618]
### Expected behavior and actual behavior

I was expecting process resubmission when I change the used variables in the nextflow.config file. However, when I use shell block, it loads from the cache. 

### Steps to reproduce the problem
```
nextflow.config file:

nextflow.enable.dsl=1
params.var1 = "a"
```

```
nextflow.nf file:

process test {
shell:
test_param = params.var1
'''
echo !{test_param}
'''
}
```

### Program output 

```
$ nextflow nextflow.nf 
N E X T F L O W  ~  version 22.10.6
Launching `nextflow.nf` [determined_solvay] DSL1 - revision: 5bafbb9a8e
executor >  local (1)
[9f/74145c] process > test [100%] 1 of 1 ✔

$ nextflow.nf -resume
N E X T F L O W  ~  version 22.10.6
Launching `nextflow.nf` [agitated_wozniak] DSL1 - revision: 5bafbb9a8e
[9f/74145c] process > test [100%] 1 of 1, cached: 1 ✔

# using vi to change params.var1 = "a" to params.var1 = "b"
vi nextflow.config 

$ nextflow nextflow.nf -resume
N E X T F L O W  ~  version 22.10.6
Launching `nextflow.nf` [sharp_euclid] DSL1 - revision: 5bafbb9a8e
[9f/74145c] process > test [100%] 1 of 1, cached: 1 ✔
```

### Environment 

* Nextflow version: 22.10.6 or 22.10.4
* Java version: 18.0.2.1
* Operating system: Linux
* Bash version: GNU bash, version 4.4.20(1)-release (x86_64-redhat-linux-gnu)

### Additional context

If I use `script:` block or define the variables in nextflow.nf file it is working as expected.

**Comment by onuryukselen** [comment]
Hello, did you able to replicate the issue?

**Comment by onuryukselen** [comment]
I replicated the same issue with DSL2.
``` 
  Version: 23.04.2 build 5870
  Created: 08-06-2023 08:29 UTC 
  System: Linux 6.2.0-1012-gcp
  Runtime: Groovy 3.0.16 on OpenJDK 64-Bit Server VM 17.0.8.1+1-Ubuntu-0ubuntu122.04
  Encoding: UTF-8 (UTF-8)
 ```

**Comment by bentsherman** [comment]
I was able to reproduce it with the following script (updated to DSL2, using latest edge release):
```groovy
params.var1 = 'a'

process test {
  shell:
  test_param = params.var1
  '''
  echo !{test_param}
  '''
}

workflow {
  test()
}
```

```bash
nextflow run main.nf

# should cache
nextflow run main.nf -resume

# shouldn't cache but caches anyway
nextflow run main.nf -resume --var1 b
```

If I refactor the shell block to a script block, it behaves correctly.

Also, with a script block I can change `test_param` to `def test_param`, which makes it a local variable instead of a global variable. But when I try this with the shell block, the run fails. Not sure if it's related, but I thought that would fix the issue.

I suspect the value of `test_param` is not being included in the task hash, although it should implicitly through the rendered shell script.

**Comment by bentsherman** [comment]
Indeed, this is the issue: https://github.com/nextflow-io/nextflow/blob/79c677a12d82400444fb6221219ce92b7247dcf2/modules/nextflow/src/main/groovy/nextflow/processor/TaskRun.groovy#L853-L857

In a script block `test_param` is a global variable, but in a shell block it is a local variable, so it is not included in the cache.

Solution could be to include the local var if the task source is a shell block. Although I wonder why it is considered a local var in the first place, I believe the same rendering engine is used for script/shell blocks just with a different escape character for variables.
